### DISCUSSION TITLE
How to access request body in the EssentialFilter?

### DISCUSSION THREAD

**Discussion by anishb266** [playframework/playframework#12275] [General]
Please help. I'm trying to access the request body via EssentialFilter. 

But the online documentation didn't show how to access it in the code.

Link - https://www.playframework.com/documentation/2.9.x/ScalaHttpFilters.

If someone can show an example of how to get the request body in the Essential Filter will be helpful.

**Comment by ihostage** [comment]
Hi, @anishb266! 
Please, read this doc part [Action composition in interaction with body parsing](https://www.playframework.com/documentation/2.9.x/ScalaActionsComposition#Action-composition-in-interaction-with-body-parsing). I think it can be useful for you.
More links for current context:
https://www.playframework.com/documentation/2.9.x/Highlights29#Deferred-Body-Parsing
https://www.playframework.com/documentation/2.9.x/ScalaHttpFilters#Filters-vs-action-composition
https://github.com/playframework/playframework/issues/9634

**Reply by anishb266** [reply]
I don't want to use action composition. I just want to use filter.

Currently, I have a filter like this given below. It's my organization code. So, for that reason I'm not exposing the entire details of the code but I can say that we are extracting header data and printing it in the logs for debugging purposes.

```
class RequestLoggingFilter @Inject()(ec: ExecutionContext) extends Filter {

  def apply(nextFilter: RequestHeader => Future[Result])(requestHeader: RequestHeader): Future[Result] = {
    logger.info("some message")
    nextFilter(requestHeader)
  }
}
```

Can you tell me how do I change the above filter to extract request body and print it?

**Comment by mkurz** [comment]
@anishb266 What is your goal / use case? Why you want to access the body in a filter? Maybe a filter is not what you want in the end.

**Reply by anishb266** [reply]
My goal is to log request body on the every endpoint call.

**Reply by anishb266** [reply]
I already have a filter in place which is logging the details. Just that it doesn't have request body in the log and I want to add it for debugging purposes.

**Comment by slisaasquatch** [comment]
Not really an answer to your question, but for Play Java, there's this thing called [ActionCreator](https://www.playframework.com/documentation/3.0.x/JavaActionCreator), where you have access to the request body like this:

```java
import com.google.inject.Inject;
import play.http.DefaultActionCreator;
import play.mvc.Action;
import play.mvc.Http;
import play.mvc.Result;

import java.lang.reflect.Method;
import java.util.concurrent.CompletionStage;

public class ExampleActionCreator extends DefaultActionCreator {

	@Inject
	public ExampleActionCreator() {}

	@Override
	public Action.Simple createAction(Http.Request request, Method actionMethod) {
		return new Action.Simple() {
			@Override
			public CompletionStage<Result> call(Http.Request req) {
				req.body(); // <------------
				return delegate.call(req);
			}
		};
	}
}
```

Unfortunately you simply do not have access to the request body in a `Filter`, and the concept of `ActionCreator` seems to be missing for Scala.
